Index: libavcodec/x86/dsputil_mmx.c
===================================================================
--- libavcodec/x86/dsputil_mmx.c	(revision 20545)
+++ libavcodec/x86/dsputil_mmx.c	(working copy)
@@ -3000,7 +3000,12 @@
         if(mm_flags & FF_MM_SSE2){
             c->int32_to_float_fmul_scalar = int32_to_float_fmul_scalar_sse2;
             c->float_to_int16 = float_to_int16_sse2;
-            c->float_to_int16_interleave = float_to_int16_interleave_sse2;
+	    /*  
+		Blender quick fix: segfaults with AC3-files. Please remove,
+		if you find out why. Probably alignment issue? (Peter Schlaile)
+
+		c->float_to_int16_interleave = float_to_int16_interleave_sse2; 
+	    */
             c->add_int16 = add_int16_sse2;
             c->sub_int16 = sub_int16_sse2;
             c->scalarproduct_int16 = scalarproduct_int16_sse2;
