Index: swscale.c
===================================================================
--- swscale.c	(revision 12850)
+++ swscale.c	(working copy)
@@ -1678,7 +1678,7 @@
 
     if(conv)
     {
-        if (dstStride[0]*srcBpp == srcStride[0]*dstBpp)
+        if (dstStride[0]*srcBpp == srcStride[0]*dstBpp && srcStride[0] > 0)
             conv(src[0], dst[0] + dstStride[0]*srcSliceY, srcSliceH*srcStride[0]);
         else
         {
